src/Account.sol

```diff
@@ -910,6 +960,127 @@ contract Account is IAccount, Auth, OpsReady {
+    /// (i.e. sUSD -> non-whitelisted token -> whitelisted token)
+    /// @param _amountIn: amount of token to swap
+    /// @param _amountOutMin: minimum amount of token to receive
+    /// @param _path: path of tokens to swap (token0 - fee - token1)
+    function _uniswapV3Swap(
+        uint256 _amountIn,
+        uint256 _amountOutMin,
+        bytes calldata _path
+    ) internal {
+        // decode tokens to swap
+        (address tokenIn, address tokenOut) = _getTokenInTokenOut(_path);
+
+        // define recipient of swap; set later once direction is established
+        address recipient;
+
+        /// @dev verify direction and validity of swap (i.e. sUSD <-> whitelisted token)
+        if (
+            tokenIn == address(MARGIN_ASSET)
+                && SETTINGS.whitelistedTokens(tokenOut)
+        ) {
+            // if swapping sUSD for another token, ensure sufficient margin
+            /// @dev margin is being transferred out of this contract
+            _sufficientMargin(int256(_amountIn));
+
+            recipient = msg.sender;
+        } else if (
+            tokenOut == address(MARGIN_ASSET)
+                && SETTINGS.whitelistedTokens(tokenIn)
+        ) {
+            // if swapping another token for sUSD, token must be transferred to this contract
+            /// @dev msg.sender must have approved Permit2 to spend at least the amountIn
+            PERMIT2.transferFrom({
+                from: msg.sender,
+                to: address(this),
+                amount: _amountIn.toUint160(),
+                token: tokenIn
+            });
+
+            recipient = address(this);
+        } else {
+            // only allow sUSD <-> whitelisted token swaps
+            revert TokenSwapNotAllowed(tokenIn, tokenOut);
+        }
+
+        // approve Permit2 to spend _amountIn of this contract's tokenIn
+        IERC20(tokenIn).approve(address(PERMIT2), _amountIn);
+
+        // approve tokens to be swapped via Universal Router
+        PERMIT2.approve({
+            token: tokenIn,
+            spender: address(UNISWAP_UNIVERSAL_ROUTER),
+            amount: _amountIn.toUint160(),
+            // Use an unlimited expiration because it most
+            // closely mimics how a standard approval works
+            expiration: type(uint48).max
+        });
+
+        _universalRouterExecute(recipient, _amountIn, _amountOutMin, _path);
+
+        EVENTS.emitUniswapV3Swap({
+            tokenIn: tokenIn,
+            tokenOut: tokenOut,
+            recipient: recipient,
+            amountIn: _amountIn,
+            amountOutMinimum: _amountOutMin
+        });
+    }
+
+    /// @notice decode and return tokens encoded in the provided path
+    /// @param _path: path of tokens to swap (token0 - fee - token1)
+    /// @return tokenIn token swapped into the respective pool
+    /// @return tokenOut token swapped out of the respective pool
+    function _getTokenInTokenOut(bytes calldata _path)
```

Comment:
**Suggestion:** This function could be done in `O(1)` by just grabbing the final 20 bytes of the `_path` to get the last address:
```solidity
    function _getTokenInTokenOut(bytes calldata _path)
        internal
        view
        returns (address tokenIn, address tokenOut)
    {
        bytes20 _tokenOut;
        assembly {
            let offset := sub(add(_path.offset, _path.length), 20)
            _tokenOut := calldataload(offset)
        }
        tokenOut = address(_tokenOut);
        tokenIn = _path.decodeFirstToken();
    }
```

Reply:
I actually had something very similar to this at one point. but in the back of my mind, I wondered why Uniswap did not implement something similar. To keep things simple, I decided to just rely on their "battle-tested" library instead of going with a solution like ^

They use something quite similar to the O(n) approach I have now for executing multi-pool swaps, so I feel even more comfortable with it (at the expense of unnecessary looping)

Reply:
Yeah fair enough, security is more important than efficiency and messing around with assembly introduces extra risk.